Python/ceval.c

```diff
@@ -920,6 +967,58 @@ _PyEval_EvalFrameDefault(PyFrameObject *f, int throwflag)
         Py_XDECREF(traceback); \
     } while(0)
 
+    /* macros for opcode cache */
+#define OPCODE_CACHE_CHECK() \
+    do { \
+        co_opt = NULL; \
+        if (co->co_opt != NULL) { \
+            unsigned char co_opt_offset = \
+                co->co_opt_opcodemap[next_instr - first_instr]; \
+            if (co_opt_offset > 0) { \
+                assert(co_opt_offset <= co->co_opt_size); \
+                co_opt = &co->co_opt[co_opt_offset - 1]; \
+                assert(co_opt != NULL); \
+                if (co_opt->optimized < 0) { \
+                    co_opt = NULL; \
+                } \
+            } \
+        } \
+    } while (0)
+
+#define OPCODE_CACHE_DEOPT() \
+    do { \
+        if (co_opt != NULL) { \
+            co_opt->optimized = -1; \
+            co->co_opt_opcodemap[next_instr - first_instr] = 0; \
```

Comment:
If you're setting the index to 0 (empty), it means you can't reopt afterwards, since the original index for this opcode is lost. Is that right?

Reply:
Yes, but it is not used for LOAD_GLOBAL cache.  I remove this for now.